nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_constructor.rs

```diff
@@ -112,12 +116,35 @@ impl StringConstructor {
         Ok(s.into_value())
     }
 
+    /// ### [22.1.2.1 String.fromCharCode ( ...`codeUnits` )](https://262.ecma-international.org/15.0/index.html#sec-string.fromcharcode)
+    ///
+    /// > This function may be called with any number of arguments which form
+    /// the rest parameter `codeUnits`.
     fn from_char_code(
-        _agent: &mut Agent,
+        agent: &mut Agent,
         _this_value: Value,
-        _arguments: ArgumentsList,
+        code_units: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // 1. Let result be the empty String.
+        if code_units.is_empty() {
+            return Ok(String::EMPTY_STRING.into_value());
+        }
+        let mut result: Vec<u8> = Vec::with_capacity(code_units.len() * 2);
+
+        // 2. For each element next of codeUnits, do
+        //   a. Let nextCU be the code unit whose numeric value is ℝ(? ToUint16(next)).
+        //   b. Set result to the string-concatenation of result and nextCU.
+        for next in code_units.iter() {
+            let code_unit = next.to_uint16(agent)?;
+            result.extend_from_slice(&code_unit.to_ne_bytes());
+        }
+
+        // 3. Return result.
+        let result_str = str::from_utf8(result.as_slice()).unwrap();
+        let result = SmallString::try_from(result_str)
```

Comment:
needs (nonblocking): `String::from_str` always allocates strings on the heap. It should attempt to inline strings that are small enough.

Reply:
Additionally, `SmallString` should expose a `MAX_LEN` constant. We could then remove all the `8` magic numbers floating around the codebase.

Reply:
It's a bit convoluted but `from_str` does indeed check and return a `SmallString` if it is possible.

A `MAX_LEN` constant would definitely be a good thing, although we actually have only _one_ place where we still check for `< 8` outside of the `small_string.rs` itself. Having the constant be abstracted would make it possible (in an easier way anyway) to generalise `SmallString` to eg. 3 bytes.